update ansible comparison statements (#48)
arcalot/arcalot.github.io: docs/arcaflow/index.md

    - It can run things on remote systems out of the box
    - It runs tasks as sequential (and, with some work, parallel) steps
    - You can pass data between steps
    - Plugins and modules can be written in a variety of languages
    - Tasks can run inside containers using custom connection plugins ([k8s example](https://docs.ansible.com/ansible/latest/collections/kubernetes/core/kubectl_connection.html))

    **What it does better than Arcaflow**
    
    - Well established and well known
    - Wide range of available plugins
    - Simple workflows are easy to write
    - It runs comands over SSH by default

>     - Users can define their own connection plugins

Reviewer: I don't think this is a unique advantage of Ansible over Arcaflow for this section. Arcaflow also uses a modular system to define connections, known as deployers.

Author: moved to section of what they do similarly

    - Users can define their own connection plugins
    
    **What it does worse than Arcaflow**
    
    - Parallelization is hard, it is intended for resource creation rather than generic workflows
    - It does not have a typing system and cannot guarantee type-safety or validation of data
    - Plugins can only be written in Python

>     - Plugins are not transparently runnable in a container engine as Ansible intends to run commands over SSH

Reviewer: I think this still applies, though perhaps it needs some adjustment to the language. Since most Ansible modules are in fact Python-based and designed to run over SSH, I don't believe those modules can simply be re-used via the k8s connection transparently, can they?

Author: It depends on what language the module is written in. If the module is written in python and there is a python interpreter in the container and the k8s connection is used, works fine.

Added statement to this effect and link to anisble docs

    - It can run things on remote systems out of the box
    - It runs tasks as sequential (and, with some work, parallel) steps
    - You can pass data between steps
    - Plugins and modules can be written in a variety of languages
    - Users can define their own connection plugins
    - Tasks can run inside containers using custom connection plugins ([k8s example](https://docs.ansible.com/ansible/latest/collections/kubernetes/core/kubectl_connection.html))

    **What it does better than Arcaflow**
    
    - Well established and well known
    - Wide range of available plugins
    - Simple workflows are easy to write
    - It runs comands over SSH by design
    - It runs comands over SSH by default
    
    **What it does worse than Arcaflow**
    
    - Parallelization is hard, it is intended for resource creation rather than generic workflows
    - It does not have a typing system and cannot guarantee type-safety or validation of data

>     - Plugins can only be written in Python

Reviewer: For Arcaflow, Python and Golang are the languages for which we currently provide an SDK. However, a plugin can be written in any language so long as that plugin adheres to the protocol and typing standards.

>     - By default, input/output data is not strongly typed. Any data type validation must be done explicitly within playbooks.

Reviewer 2: I think one point that is less obvious with this wording is the type validation doesn't happen until the step is run, whereas with the type system in arcaflow, it can validate it without executing the steps.

This is a critical point. It's like using Python/Javascript vs Java/C++/Rust/TypeScript.

>     - By default, input/output data is not strongly typed. Any data type validation must be done explicitly within playbooks.

Author: ```suggestion
    - By default, input/output data is not strongly typed. Any data type validation must be done explicitly within playbooks, and is executed at run time.
```

>     - It is written in Python, so the control host where it is run from needs system python dependencies (or to run from inside a container or python virtualenv)

Author: ```suggestion
    - It is written in Python, so the control host where it is run from needs system python dependencies.
        - Alternatively, you can run it from inside a container or python virtualenv
```

    - Users can define their own connection plugins
    - Tasks can run inside containers using custom connection plugins ([k8s example](https://docs.ansible.com/ansible/latest/collections/kubernetes/core/kubectl_connection.html))

    **What it does better than Arcaflow**
    
    - Well established and well known
    - Wide range of available plugins
    - Simple workflows are easy to write
    - It runs comands over SSH by default
    
    **What it does worse than Arcaflow**
    
    - Ansible's approach to parallelization is in terms of executing the same tasks against different hosts in parallel (see "forks" and "strategy"). Defining different tasks to perform in parallel is more challenging.
    - By default, input/output data is not strongly typed. Any data type validation must be done explicitly within playbooks, and is executed at run time.
    - It is written in Python, so the control host where it is run from needs system python dependencies.
        - Alternatively, you can run it from inside a container or python virtualenv
    - Depending on the module used, there may be system requirements for python or other dependencies on the target host/container.

>         -  See [ansible documentation](https://docs.ansible.com/ansible/latest/reference_appendices/faq.html#how-do-i-handle-not-having-a-python-interpreter-at-usr-bin-python-on-a-remote-machine)

Reviewer 3: Sorry for the random review, I was pointed to this PR and thought I would weigh in.

I don't personally see these as `**What it does worse than Arcaflow**`. 

Quite frankly I find it troubling that there is this measuring stick approach to other open source projects. Why the need to down other projects to prop this project up? 

What if your binary is built and only works on x86-64, since the engine is in Go (to argue this whole beef with python). Sure, someone could spend the time to build a ARM or Power binary, similar they could just install the expected dependencies.

Author: I was sticking with the headings as they were given, I agree most software will have some requirements on the host it is running :D so its bit hard to say which requirements are worse than others

Reviewer 3: If we want to compare, might I suggest, instead of suggesting one way is better or worse

- `Here is what Arca does` (list the things)
- `Here is what Ansible does` (list the things)

Or even better, not even mention other projects here.

If the discussion is more about "This might overlap with that" or "This project has duplicate functionality", I think we should look at any issues/rfe's the team here has opened against existing open source projects (like Ansible, Ariflow, and others this).  

If there were no opened issues or rfes against existing projects, then 🤷🏽

Reviewer 2: I'd consider this more informative than downing other projects. We list pros and cons, not just cons, with some bias towards our project.
It's a simple guide to figure out the right tool for your job.

Reviewer 3: @jaredoconnell ack - Thanks! I still wonder if anyone open up issues against these projects (like Ansible and how they handle parallelism) to maybe propose solutions to the `cons` you have mentioned here?

Reviewer: Please remember to assume positive intent. The reason we have these sections in the docs is because we were fielding questions about when Arcaflow was or was not the best solution versus these and other alternatives. So this is a sort of FAQ designed to both steer people toward and away from Arcaflow as appropriate.

It seems like the concern may be about the tone implied with "does worse" and "does better" language. I think we should consider how to better label these things to achieve the goal of informing users while avoiding a possible perception of a negative tone or judgment.

Reviewer: I've submitted issue #52 to address the observed tonal concerns separately.

Reviewer 3: ack @dustinblack so there are no open issues/PRs against these other projects to help improve them. 

Thanks for #52!

Reviewer 3: ack - interesting comparison. 

in my opinion it is less about calling out flaws, but improving collaboration. 

Maybe there are things the motorcycle manufactures could improve on, but they doesn't see it because they are too close to building motorcycles. 

Thanks for your perspective!

    - Wide range of available plugins
    - Simple workflows are easy to write
    - It runs comands over SSH by default
    
    **What it does worse than Arcaflow**
    
    - Ansible's approach to parallelization is in terms of executing the same tasks against different hosts in parallel (see "forks" and "strategy"). Defining different tasks to perform in parallel is more challenging.
    - By default, input/output data is not strongly typed. Any data type validation must be done explicitly within playbooks.
    - It is written in Python, so the control host where it is run from needs system python dependencies (or to run from inside a container or python virtualenv)

>     - Depending on the module used, there may be system requirements for python or other dependencies on the target host/container. See [ansible documentation](https://docs.ansible.com/ansible/latest/reference_appendices/faq.html#how-do-i-handle-not-having-a-python-interpreter-at-usr-bin-python-on-a-remote-machine)

Reviewer: I think I am fine with the content changes, but my nit is that this strays from the style of the document. All of the other comparison bullet points are succinct one-liners, and these are sentences with qualifications and references. I wonder if we can address the factual and tonal concerns here without the need for defensive posture.

Author: ```suggestion
    - Depending on the module used, there may be system requirements for python or other dependencies on the target host/container.
        -  See [ansible documentation](https://docs.ansible.com/ansible/latest/reference_appendices/faq.html#how-do-i-handle-not-having-a-python-interpreter-at-usr-bin-python-on-a-remote-machine)
```
